trio/_channel.py

```diff
@@ -68,36 +97,56 @@ def open_memory_channel(max_buffer_size):
         raise TypeError("max_buffer_size must be an integer or math.inf")
     if max_buffer_size < 0:
         raise ValueError("max_buffer_size must be >= 0")
-    state = MemoryChannelState(max_buffer_size)
+    state: MemoryChannelState[T] = MemoryChannelState(max_buffer_size)
     return (
-        MemorySendChannel._create(state),
-        MemoryReceiveChannel._create(state),
+        MemorySendChannel[T]._create(state),
+        MemoryReceiveChannel[T]._create(state),
     )
 
 
+# This workaround requires python3.9+, once older python versions are not supported
+# or there's a better way of achieving type-checking on a generic factory function,
+# it could replace the normal function header
+if TYPE_CHECKING:
+    # written as a class so you can say open_memory_channel[int](5)
+    class open_memory_channel(tuple[MemorySendChannel[T], MemoryReceiveChannel[T]]):
```

Comment:
CI is complaining about this, heh. We can just `type: ignore` it because we know CI is wrong here (it's not run at runtime).

Reply:
That's because the check is run on 3.8 ... which seems kinda old by this point.
It's easily fixed by using `Tuple`, so I'll just do that until the check is run on newer stuff.